---
title: "Phyloseq"
author: "Maggi Mars & Sasha Mikheyev"
date: "4/17/2018"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, cache = TRUE)
```

```{r loadLibraries}
library("phyloseq")
library("ggplot2")
theme_set(theme_bw())
library("dplyr")
library("tidyr")
library("RColorBrewer")
library(gridExtra)
library(vegan)
library(DESeq2)
library(plyr)
```

```{r abundance}
P <- brewer.pal(12, "Paired")
S2 <- brewer.pal(8, "Set2")
S1 <- brewer.pal(8, "Set1")
D2 <- brewer.pal(8, "Dark2")
colors<- c(P, S2, S1,D2)
colors2<- c(P, S2, S1,D2)

#import taxonomy (saved as tsv from taxonomy.qzv)
taxonomy<- read.delim("Taxonomy.tsv")
names(taxonomy) <- c("row", "tax", "Confidence")
row.names(taxonomy) <-taxonomy[[1]]
taxonomy <- taxonomy[,(-1)]

taxonomy <-  separate(taxonomy, tax, c("D0","D1", "D2", "D3", "D4", "D5", "D6", "D7", "D8", "D9", "D10", "D11", "D12", "D13", "D14"), sep = ";", fill = "right")
taxonomy <- taxonomy[,c(1:6)]
taxmat <- as.matrix(taxonomy)

TAX = tax_table(taxmat)

#import metadata
metatable <- read.delim("Metadata2.txt")
#View(metatable)
row.names(metatable) <- metatable[[1]]
metatable<- metatable[,(-1)]

META <- sample_data(metatable)

#Import Phylogenetic tree (rooted, exported from qiime2)
TREE<- read_tree("tree.nwk")

#Import Feature Table
Svtab <- read.delim("feature-table.txt")
row.names(Svtab)<-Svtab[[1]] #make OTU ID the row names
Svtab<-Svtab[,-(1)] # remove OTU ID column
fmat <- as.matrix(Svtab)
OTU = otu_table(fmat, taxa_are_rows = TRUE)

#create phyloseq object
ps<- phyloseq(OTU, TAX, META, TREE)

#subset to only include GUT samples (exclude media samples)

ps<- subset_samples(ps, SampleType=="Gut")

rank_names(ps)
table(tax_table(ps)[, "D0"], exclude = NULL)
#remove sequences that have D0 unassigned
# Filter Physeq object - Bacteria ONLY
ps <- subset_taxa(ps, D0=="D_0__Bacteria")
table(tax_table(ps)[, "D2"], exclude = NULL)
#remove phyla that cannot be part of microbiome
filterPhyla = c("D_2__Chloroplast" , " D_2__Cyanobacteria")
ps = subset_taxa(ps, !D2 %in% filterPhyla)

prevdf = apply(X = otu_table(ps),
               MARGIN = ifelse(taxa_are_rows(ps), yes = 1, no = 2),
               FUN = function(x){sum(x > 0)})

prevdf = data.frame(Prevalence = prevdf,
                    TotalAbundance = taxa_sums(ps),
                    tax_table(ps))

plyr::ddply(prevdf, "D2", function(df1){cbind(mean(df1$Prevalence),sum(df1$Prevalence))})

ggplot(prevdf, aes(TotalAbundance, Prevalence / nsamples(ps),color=D2)) +
  # Include a guess for parameter
  geom_hline(yintercept = 0.01, alpha = 0.5, linetype = 2) +  geom_point(size = 2, alpha = 0.7) +
  scale_x_log10() +  xlab("Total Abundance") + ylab("Prevalence [Frac. Samples]") +
  facet_wrap(~D2) + theme(legend.position="none")

prevalenceThreshold = 0.1 * nsamples(ps) # remove OTS found in less than 10% of the samples
prevalenceThreshold

keepTaxa = rownames(prevdf)[(prevdf$Prevalence >= prevalenceThreshold)]
ps2 = prune_taxa(keepTaxa, ps)

table(tax_table(ps2)[, "D2"], exclude = NULL)

prevdf2 = apply(X = otu_table(ps2),
               MARGIN = ifelse(taxa_are_rows(ps2), yes = 1, no = 2),
               FUN = function(x){sum(x > 0)})

prevdf2 = data.frame(Prevalence = prevdf2,
                    TotalAbundance = taxa_sums(ps2),
                    tax_table(ps2))

plyr::ddply(prevdf2, "D1", function(df1){cbind(mean(df1$Prevalence),sum(df1$Prevalence))})

ggplot(prevdf2, aes(TotalAbundance, Prevalence / nsamples(ps),color=D2)) +
  # Include a guess for parameter
  geom_hline(yintercept = 0.01, alpha = 0.5, linetype = 2) +  geom_point(size = 2, alpha = 0.7) +
  scale_x_log10() +  xlab("Total Abundance") + ylab("Prevalence [Frac. Samples]") +
  facet_wrap(~D1) + theme(legend.position="none")

#transform to relative abundance
ps2ra = transform_sample_counts(ps2, function(x){x / sum(x)})

#subset by generation
ps2raF0 <- subset_samples(ps2ra, Generation=="F0")
ps2raF1 <- subset_samples(ps2ra, Generation=="F1")

#Relabundance Plots
taxabarplot<-plot_bar(ps2ra, fill = "D2") + scale_y_continuous(expand = c(0, 0)) + ggtitle("Level 2") + scale_fill_manual(values=colors2) + theme(legend.title=element_blank()) + geom_bar(aes( fill=D2), stat="identity", position="stack")

taxabarplot + theme(legend.position="none")

#legend
g_legend<-function(a.gplot){
  tmp <- ggplot_gtable(ggplot_build(a.gplot))
  leg <- which(sapply(tmp$grobs, function(x) x$name) == "guide-box")
  legend <- tmp$grobs[[leg]]
  legend
}
legend <- g_legend(taxabarplot)
SVlegend <- grid.arrange(legend)

##################################
#Remove all taxa that are D1 == NA

filterPhyla = c("D_1__Acidobacteria", "D_1__Actinobacteria", "D_1__Bacteroidetes", "D_1__Firmicutes", " D_1__Parcubacteria", "D_1__Proteobacteria", "D_1__Verrucomicrobia")
ps3 = subset_taxa(ps2, D1 %in% filterPhyla)
table(tax_table(ps3)[, "D1"], exclude = NULL)

#Final Prevalence Plot
prevdf3 = apply(X = otu_table(ps3),
                MARGIN = ifelse(taxa_are_rows(ps3), yes = 1, no = 2),
                FUN = function(x){sum(x > 0)})

prevdf3 = data.frame(Prevalence = prevdf3,
                     TotalAbundance = taxa_sums(ps3),
                     tax_table(ps3))

plyr::ddply(prevdf3, "D1", function(df1){cbind(mean(df1$Prevalence),sum(df1$Prevalence))})

ggplot(prevdf3, aes(TotalAbundance, Prevalence / nsamples(ps),color=D2)) +
  # Include a guess for parameter
  geom_hline(yintercept = 0.01, alpha = 0.5, linetype = 2) +  geom_point(size = 2, alpha = 0.7) +
  scale_x_log10() +  xlab("Total Abundance") + ylab("Prevalence [Frac. Samples]") +
  facet_wrap(~D2) + theme(legend.position="none")

#transform to relative abundance
ps3ra = transform_sample_counts(ps3, function(x){x / sum(x)})

#subset by generation
ps3raF0 <- subset_samples(ps3ra, Generation=="F0")
ps3raF1 <- subset_samples(ps3ra, Generation=="F1")

#Relabundance Plots
filterGens = c("F0", "F1")
ps4 <- subset_samples(ps3ra, Generation %in% filterGens)
taxabarplot<-plot_bar(ps4, fill = "D2") + scale_y_continuous(expand = c(0, 0)) + ggtitle("F0 & F1 - Level 2 Taxonomy") + scale_fill_manual(values=colors2) + theme(legend.title=element_blank()) + geom_bar(aes( fill=D2), stat="identity", position="stack")

taxabarplot + theme(legend.position="none")

#legend
g_legend<-function(a.gplot){
  tmp <- ggplot_gtable(ggplot_build(a.gplot))
  leg <- which(sapply(tmp$grobs, function(x) x$name) == "guide-box")
  legend <- tmp$grobs[[leg]]
  legend
}
legend <- g_legend(taxabarplot)
SVlegend <- grid.arrange(legend)
```

```{r adonis}
#F0
OTUs <- t(data.frame(otu_table(ps3raF0))) #get data frame of OTUs from phyloseq object
meta <- metatable[metatable$Generation=="F0",]
meta <- meta[meta$SampleType=="Gut",]

adonis(vegdist(OTUs, method = "bray") ~ Treatment, data = meta)

#pairwise CD ~ NS F0
CDNS0 <- subset_samples(ps3raF0, Treatment %in% c("CD" , "NSD"))
OTUs <- t(data.frame(otu_table(CDNS0))) #get data frame of OTUs from phyloseq object
metaPW <- meta[meta$Treatment %in% c("NSD", "CD"),]
adonis(vegdist(OTUs, method = "bray") ~ Treatment, data = metaPW)


#F1
OTUs <- t(data.frame(otu_table(ps3raF1))) #get data frame of OTUs from phyloseq object
meta <- metatable[metatable$Generation=="F1",]
meta <- meta[meta$SampleType=="Gut",]
adonis(vegdist(OTUs, method = "bray") ~ Treatment, data = meta)

#pairwise CD ~ NS F1
CDNS1 <- subset_samples(ps3raF1, Treatment %in% c("CD" , "NSD"))
OTUs <- t(data.frame(otu_table(CDNS1))) #get data frame of OTUs from phyloseq object
metaPW <- meta[meta$Treatment %in% c("NSD", "CD"),]
adonis(vegdist(OTUs, method = "bray") ~ Treatment, data = metaPW)
```

## Analyze responses using phyloseq

```{r phyloseq}
testdiets <- c("CD", "NSD")
ps3cn <- subset_samples(ps3, Treatment %in% testdiets)
ps3F01 <- subset_samples(ps3cn, Generation=="F0" | Generation=="F1")
ddse <- phyloseq_to_deseq2(ps3F01, ~Treatment+Generation+ Treatment:Generation)
ddse2 <- DESeq(ddse, test="Wald", fitType="local")
resultsNames(ddse2)
res <- results(ddse2,  name = "TreatmentNSD.GenerationF1")

alpha <-  0.05
sigtab <- res[which(res$padj < alpha), ]

taxCounts <- list()
i <- 1
for (taxon in rownames(sigtab)) {
  taxName <- as.data.frame(tax_table(ps3F01)[taxon])[,6]
  print(taxName)
  dTrt <- plotCounts(ddse2, gene=taxon, intgroup="Treatment",returnData=TRUE)
  dGen <- plotCounts(ddse2, gene=taxon, intgroup="Generation",returnData=TRUE)
  tc <- cbind(dTrt,dGen)[,-3]
  tc$species <- taxName
  taxCounts[[i]] <- tc
  i <- i + 1
}

ggplot(rbind.fill(taxCounts), aes(x=Treatment, y=count, color=Treatment))+ geom_boxplot(position=position_dodge()) + scale_y_log10(breaks=c(25,100,400))+facet_grid(species~Generation, scale = "free_y")
```